shared/middleware/files_api.rb

```diff
@@ -404,6 +404,18 @@ def put_file(endpoint, encrypted_channel_id, filename, body)
       return bad_request if share_failure && share_failure[:type] != "address"
     end
 
+    # Don't allow project to be saved if it contains non-UTF-8 characters (causing error / project to not load when opened).
+    if 'sources' == endpoint
+      body_json = JSON.parse(body)
+      source = body_json["source"]
+      html = body_json["html"]
+
+      source_is_valid = source && has_valid_encoding?(source)
+      html_is_valid = html ? source.force_encoding("UTF-8").valid_encoding? : true # HTML only exists for AppLab projects
+
+      return bad_request unless source_is_valid && html_is_valid
+    end
```

Comment:
`test_files.rb` would be a good place to add a test that checks that this function identifies non-utf8 when needed.

Reply:
good idea! Any ideas on how to generate non-utf8 characters? haha I feel like I maybe tried but couldn't figure anything out?

Reply:
Yeah, it seems hard. I wondered if other folks had run into this and this thread seems to have some discussion on it. https://stackoverflow.com/questions/15343781/testing-of-graceful-handling-of-invalid-utf-8-in-rails. There's also this very tiny gist. https://gist.github.com/kamarcum/3969065 Were these methods you've tried already?

Reply:
Ooo I haven't seen the gist I'll try that!